backends/utils/history_server.go

```diff
@@ -0,0 +1,363 @@
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to get '%v'", currentFilePath)
+		}
+
+		if !fileIterator.Next() {
+			if fileIterator.Error() != nil {
+				return nil, err
+			}
+			continue
+		}
+
+	ScanningFileChunk:
+		scanner := bufio.NewScanner(bytes.NewReader(fileIterator.At()))
+		for scanner.Scan() {
+			var currentRow []HistoryEntry
+
+			if err := json.Unmarshal([]byte(scanner.Text()), &currentRow); err != nil {
+				// Possibly reached end of file chunk, get next chunk and retry marshaling
+				if fileIterator.Next() {
+					scanner = bufio.NewScanner(bytes.NewReader(append(scanner.Bytes(), fileIterator.At()...)))
+					if scanner.Scan() {
+						if innerErr := json.Unmarshal([]byte(scanner.Text()), &currentRow); innerErr != nil {
+							return nil, innerErr
+						}
+					} else {
+						return nil, scanner.Err()
+					}
+				} else {
+					return nil, fmt.Errorf("error reading logs. json marshal error:%v\n iterator error: %v", err, fileIterator.Error())
+				}
+			}
+
+			for _, entry := range currentRow {
+				// filter out logs
+				if !filter.filter(entry) {
+					continue
+				}
+				_ = AppendColumn(defaultColumns[0], entry.UserName)
+				_ = AppendColumn(defaultColumns[1], entry.BackendName)
+				_ = AppendColumn(defaultColumns[2], entry.TableName)
+				_ = AppendColumn(defaultColumns[3], entry.ActionDuration.Nanoseconds()/1e6)
+				_ = AppendColumn(defaultColumns[4], entry.StartTime)
+				_ = AppendColumn(defaultColumns[5], entry.ActionType)
+				_ = AppendColumn(defaultColumns[6], entry.Container)
+
+				for k, v := range entry.AdditionalData {
+					// If this column already exists, append new data
+					if _, ok := customColumnsByName[k]; !ok {
+						// Create a new column with nil values up until this row
+						data := make([]string, i)
+						data[i-1] = v
+						customColumnsByName[k], err = frames.NewSliceColumn(k, data)
+						if err != nil {
+							return nil, err
+						}
+					} else {
+						_ = AppendColumn(customColumnsByName[k], v)
+					}
+				}
+
+				// Add null values
+				for columnName, col := range customColumnsByName {
+					if _, ok := entry.AdditionalData[columnName]; !ok {
+						_ = AppendNil(col)
+					}
+				}
+
+				_ = AppendColumn(indexColumn, i)
+				i++
+			}
+		}
+
+		if scanner.Err() != nil {
+			return nil, err
+		}
+
+		// in case previous chunk ended exactly in the end of a row and we still got more data to process
+		if fileIterator.Next() {
+			goto ScanningFileChunk
+		} else if fileIterator.Error() != nil {
+			return nil, fileIterator.Error()
+		}
+	}
+
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+	allColumns := defaultColumns
+
+	for _, col := range customColumnsByName {
+		allColumns = append(allColumns, col)
+	}
+
+	return frames.NewFrame(allColumns, []frames.Column{indexColumn}, nil)
+}
+
+func (m *HistoryServer) writeMonitoringBatch(logs []HistoryEntry) {
+	d, err := json.Marshal(logs)
```

Comment:
Wouldn't it make more sense to have one entry per line?

Reply:
hmm yea thought about this.. maybe yes. this will potentially reduce the number of `unmarshals` that we do. you think it's better just a single entry per line?

Reply:
Yup